File: scratchpad/ui-sweeps/seed-libtext.py
"""Give the seeded uploads the readings INBOX 279's screenshots show.

`seed-libtext.js` uploads the files; this writes what a vision model would
have written, because there is none in the sandbox and the shape under test is
the rendered card, not the inference. Two things go in:

* the images get a caption, a caption model, a vision reading and its model,
  which is every rank the image card's foot can carry;
* the PDFs get fourteen `page_reads` rows each, so `pages_read` is 14 and the
  joined reading is long enough for the Files block's word count to be a real
  number rather than a handful.

    python scratchpad/ui-sweeps/seed-libtext.py /tmp/mm-dens/memorymap.db
"""

import sqlite3
import sys

db = sys.argv[1] if len(sys.argv) > 1 else "/tmp/mm-dens/memorymap.db"
con = sqlite3.connect(db)

CAPTION = (
    "A printed notice board behind glass, listing opening hours and three "
    "membership rates in a serif face"
)
OCR = (
    "Rates and opening hours\n"
    "Monday to Friday, nine until six. Saturday, ten until four.\n"
    "Day rate twenty two pounds, week rate ninety, month rate two hundred and eighty.\n"
    "Members pay the week rate for a day pass on any weekday before noon."
)

PAGE = (
    "Section {n}. The rates in this handbook apply from the first of the month "
    "and are reviewed every spring. A member who joins part way through a month "
    "pays a pro rata share of the month rate, counted in whole days from the "
    "day the form is signed rather than from the first visit. Concessions are "
    "listed overleaf and are not cumulative: where two apply, the larger one is "
    "taken and the other is recorded but not charged against."
)

images = con.execute(
    "SELECT id, original_name FROM media_uploads WHERE original_name LIKE '%.png' ORDER BY id"
).fetchall()
for index, (media_id, name) in enumerate(images):
    # The third image is left bare on purpose: a card with nothing read is the
    # short foot the design has to keep short.
    if name.startswith("plain"):
        continue
    con.execute(
        "UPDATE media_uploads SET caption = ?, caption_model = ?, vision_ocr_text = ?, "
        "vision_ocr_model = ? WHERE id = ?",
        (CAPTION, "qwen3-vl:4b", OCR, "GLM-OCR-GGUF:Q8_0", media_id),
    )
    if index == 1:
        # One card with Tesseract's reading as well, which is the second box
        # inside the fold.
        con.execute(
            "UPDATE media_uploads SET ocr_text = ? WHERE id = ?",
            ("RATES AND OPENING HOURS\nMon-Fri 9-6  Sat 10-4", media_id),
        )

files = con.execute(
    "SELECT id FROM media_uploads WHERE original_name LIKE '%.pdf' ORDER BY id"
).fetchall()
for (media_id,) in files:
    con.execute("DELETE FROM page_reads WHERE kind = 'upload' AND source_id = ?", (media_id,))
    for page in range(1, 15):
        con.execute(
            "INSERT INTO page_reads (kind, source_id, page, reader, model, text, caption, "
            "caption_model, created_at) VALUES ('upload', ?, ?, 'vision', ?, ?, '', '', "
            "datetime('now'))",
            (media_id, page, "qwen3-vl:4b", PAGE.format(n=page)),
        )

con.commit()
print(
    "images",
    len(images),
    "files",
    len(files),
    "page rows",
    con.execute("SELECT COUNT(*) FROM page_reads").fetchone()[0],
)
con.close()

File: scratchpad/ui-sweeps/seed-libtext.js
// Seed the Library with the two rows INBOX 279 is about: an image that has a
// caption AND a vision reading (so the card's foot carries every rank the
// report lists), and a PDF with a real page-by-page reading (so the Files
// sub-tab's "Text extracted from this file" block has pages, words and a
// first sentence rather than an empty state).
//
// Uploads only: the caption, the model names and the page readings are
// written by `seed-libtext.py` straight into the database, because producing
// them for real needs a vision model this sandbox does not have and the
// shapes under test are the *rendered* ones.
//
//   BASE=http://127.0.0.1:8991 PLAYWRIGHT_BROWSERS_PATH=/opt/pw-browsers \
//     NODE_PATH=/opt/node22/lib/node_modules node seed-libtext.js
const { boot } = require('./lib.js');

const SHOT = `<!doctype html><meta charset="utf-8"><body style="margin:0;width:900px;height:620px;background:#eef2f6;font-family:Georgia,serif;color:#22303c">
<div style="padding:60px 70px;line-height:1.5;font-size:28px">
<div style="font-size:40px;font-weight:bold;margin-bottom:18px">Rates and opening hours</div>
<p>Monday to Friday, nine until six. Saturday, ten until four.</p>
<p>Day rate twenty two pounds, week rate ninety, month rate two hundred and eighty.</p>
<p>Members pay the week rate for a day pass on any weekday before noon.</p>
</div></body>`;

function onePagePdf(title) {
  const body = `BT /F1 12 Tf 40 120 Td (${title}) Tj ET`;
  const objs = [
    '<</Type/Catalog/Pages 2 0 R>>',
    '<</Type/Pages/Kids[3 0 R]/Count 1>>',
    '<</Type/Page/Parent 2 0 R/MediaBox[0 0 400 400]/Resources<</Font<</F1 5 0 R>>>>/Contents 4 0 R>>',
    `<</Length ${body.length}>>\nstream\n${body}\nendstream`,
    '<</Type/Font/Subtype/Type1/BaseFont/Helvetica>>',
  ];
  let pdf = '%PDF-1.4\n';
  const offsets = [];
  objs.forEach((o, i) => { offsets.push(pdf.length); pdf += `${i + 1} 0 obj\n${o}\nendobj\n`; });
  const xref = pdf.length;
  pdf += `xref\n0 ${objs.length + 1}\n0000000000 65535 f \n`;
  for (const off of offsets) pdf += String(off).padStart(10, '0') + ' 00000 n \n';
  pdf += `trailer\n<</Size ${objs.length + 1}/Root 1 0 R>>\nstartxref\n${xref}\n%%EOF\n`;
  return pdf;
}

(async () => {
  const { browser, page } = await boot();
  const shotCtx = await browser.newContext();
  const shotPage = await shotCtx.newPage();
  await shotPage.setViewportSize({ width: 900, height: 620 });
  await shotPage.setContent(SHOT);
  const png = await shotPage.screenshot({ type: 'png' });
  await shotCtx.close();

  const out = await page.evaluate(async (args) => {
    const { b64, pdfs } = args;
    const post = async (name, blob, type) => {
      const fd = new FormData();
      fd.append('file', new File([blob], name, { type }));
      fd.append('direct', 'true');
      // Raw fetch, not `api()`: `api()` forces a JSON content type, which
      // strips the multipart boundary and makes the upload a 422.
      const r = await fetch('/media/upload', {
        method: 'POST', body: fd,
        headers: { 'X-Auth-Token': authToken(), 'X-Workspace-ID': activeSpaceId() },
      });
      return await r.json();
    };
    const bytes = Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
    const shots = [];
    for (const name of ['price-board.png', 'notice-board.png', 'plain-sky.png']) {
      shots.push(await post(name, bytes, 'image/png'));
    }
    const files = [];
    for (const pdf of pdfs) files.push(await post(pdf[0], pdf[1], 'application/pdf'));
    // One reference each for the first image and the first file, so "Used in
    // 1 place" is a real count rather than a staged string.
    const refs = [];
    //: `POST /media/upload` answers `{id, url, filename}`; `original_name` is
    //: the listing's field (`MediaUploadOut`), so reading it here wrote
    //: `![undefined](...)` into every seeded note (pass2.md, Remaining 6).
    for (const up of [shots[0], files[0]]) {
      if (up && up.url) {
        const r = await api('/entries', {
          method: 'POST',
          body: JSON.stringify({ content: `# Where this came from\n\n![${up.filename || up.original_name || 'Attached file'}](${up.url})`, tags: ['library'] }),
        });
        refs.push(r.status);
      }
    }
    return { shots: shots.map((s) => s.id), files: files.map((f) => f.id), refs };
  }, { b64: png.toString('base64'), pdfs: [['rates-handbook.pdf', onePagePdf('Rates handbook')], ['meeting-pack.pdf', onePagePdf('Meeting pack')]] });
  console.log(JSON.stringify(out));
  await browser.close();
})().catch((e) => { console.error(e); process.exit(1); });
